File: SurvCop/src/survival_nllk.f90
!program survivalnllkmain
!  implicit none
!  integer n,i,d
!  integer A(5,5), fam(5,5), temi(25)
!  double precision temi2(20), nllk, parammat1(5,5), parammat2(5,5)
!  double precision datacdfplus(4,5), datacdfminus(4,5)
!  logical censorstatus(4), isdisc(5)
!
!  n=4
!  d=5
!  temi = (/1,0,0,0,0, 1,3,0,0,0, 3,1,2,0,0, 3,2,1,4,0, 4,3,2,1,5/)
!  A = reshape(temi, (/5,5/))
!  temi = (/0,0,0,0,0, 5,0,0,0,0, 5,5,0,0,0, 5,5,5,0,0, 5,5,5,5,0/)
!  fam = reshape(temi, (/5,5/))
!  temi2 = (/.5,.4,.7,.5, .4,.6,.9,.7, .5,.3,.2,.4, .4,.4,.7,.4, .7,.4,.5,.4/)
!  datacdfplus = reshape(temi2, (/4,5/))
!  datacdfminus = reshape(temi2, (/4,5/))
!  parammat1 = 4.d0
!  parammat2 = 0.d0
!  censorstatus = (/.true., .false., .true., .false./)
!  isdisc = (/.false., .false., .false., .false., .false./)
!  call survivalnllk(n, d, parammat1, parammat2, fam, A, datacdfplus, datacdfminus, censorstatus, isdisc, nllk)
!end

subroutine survivalnllk(n, d, parammat1, parammat2, fam, A, datacdfplus, datacdfminus, censorstatus, isdisc, nllk)
  implicit none
  integer n, d, i, j, k, l
  integer fam(d,d), A(d,d), M(d,d), Imat(d,d)
  double precision parammat1(d,d), parammat2(d,d)
  double precision datacdfplus(n,d), datacdfminus(n,d)
  double precision vplus(n,d), vminus(n,d), vpplus(n,d), vpminus(n,d), splus(n,d), sminus(n,d), wplus(n,d), wminus(n,d)
  logical censorstatus(n), isdisc(d)
  double precision nllk, tmp, tmp1, tmp2, tmp3, tmp4
  !print*, n, d
  !print*, parammat1, parammat2
  !print*, fam
  !print*, (A(i,:), i=1,5)
  !print*, datacdfplus
  !print*, datacdfminus
  !print*, censorstatus
  !print*, isdisc
  !print*, nllk
  
  vplus=0.d0; vminus=0.d0; vpplus=0.d0; vpminus=0.d0; splus=0.d0; sminus=0.d0; wplus=0.d0; wminus=0.d0
  M=0; Imat = 0
  do i = 1,d
    do j = i,d
      M(i, j) = maxval(A(1:i, j))
    end do
  end do
  !print*, M
  
  do i = 2,d
    do j = i,d
      if (A(i, j) < M(i, j)) then
        Imat(i-1, M(i, j)) = 1
      end if
    end do
  end do
  !print*, Imat
  
  nllk = 0.d0
  do j = 1,d
    splus(:, j) = datacdfplus(:, A(1, j))
    sminus(:, j) = datacdfminus(:, A(1, j))
    wplus(:, j)= datacdfplus(:, j)
    wminus(:, j) = datacdfminus(:, j)
  end do
  
  do j = 2,d-1
    do k = 1,n
      call mixedcopulallk(splus(k, j), sminus(k, j), wplus(k, j), wminus(k, j), isdisc(A(1, j)), isdisc(j), fam(1, j),&
      parammat1(1, j), parammat2(1, j), tmp)
      nllk = nllk - tmp
    end do
  end do
  !call dblepr("level1", 8, nllk, 1)
  
  do k = 1,n
    if (.not. censorstatus(k)) then
      call mixedcopulallk(splus(k,d), sminus(k,d), wplus(k,d), wminus(k,d), isdisc(A(1,d)), isdisc(d), fam(1,d),&
      parammat1(1,d), parammat2(1,d), tmp)
      nllk = nllk - tmp
    end if
  end do
  !call dblepr("level1", 8, nllk, 1)
  
  do l = 2,d-1
    where (splus >= .999d0)
      splus = .999d0
    end where
    where (sminus >= .999d0)
      sminus = .998d0
    end where
    where (wplus >= .999d0)
      wplus = .999d0
    end where
    where (wminus >= .999d0)
      wminus = .998d0
    end where
    where (splus <= .001d0)
      splus = .002d0
    end where
    where (sminus <= .001d0)
      sminus = .001d0
    end where
    where (wplus <= .001d0)
      wplus = .002d0
    end where
    where (wminus <= .001d0)
      wminus = .001d0
    end where
    
    do j = 1,d
      if (Imat(l-1, j) == 1) then
        if (isdisc(j)) then
          do k = 1,n
            call bivcopcdf(splus(k,j), wplus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
            call bivcopcdf(splus(k,j), wminus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
            vpplus(k,j) = (tmp1 - tmp2) / (wplus(k,j) - wminus(k,j))
            call bivcopcdf(sminus(k,j), wplus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp3)
            call bivcopcdf(sminus(k,j), wminus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp4)
            vpminus(k,j) = (tmp3 - tmp4) / (wplus(k,j) - wminus(k,j))
          end do
        else
          do k = 1,n
            call bivcopccdf(splus(k,j), wplus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
            vpplus(k,j) = tmp1
            call bivcopccdf(sminus(k,j), wplus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
            vpminus(k,j) = tmp2
          end do
        end if
      end if
      if (isdisc(A(l-1, j))) then
        do k = 1,n
          call bivcopcdf(splus(k,j), wplus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
          call bivcopcdf(sminus(k,j), wplus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
          vplus(k,j) = (tmp1 - tmp2) / (splus(k,j) - sminus(k,j))
          call bivcopcdf(splus(k,j), wminus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp3)
          call bivcopcdf(sminus(k,j), wminus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp4)
          vminus(k,j) = (tmp3 - tmp4) / (splus(k,j) - sminus(k,j))
        end do
      else
        do k = 1,n
          call bivcopccdf(wplus(k,j), splus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp1)
          vplus(k,j) = tmp1
          call bivcopccdf(wminus(k,j), splus(k,j), fam(l-1,j), parammat1(l-1,j), parammat2(l-1,j), tmp2)
          vminus(k,j) = tmp2
        end do
      end if
    end do
    
    do j = l+1,d
      if (A(l,j) == M(l,j)) then
        splus(:, j) = vplus(:, M(l,j))
        sminus(:, j) = vminus(:, M(l,j))
      else
        splus(:, j) = vpplus(:, M(l,j))
        sminus(:, j) = vpminus(:, M(l,j))
      end if
      wplus(:, j) = vplus(:, j)
      wminus(:, j) = vminus(:, j)
    end do
    
    if (l < (d - 1)) then
      do j = l+1,d-1
        do k = 1,n
          call mixedcopulallk(splus(k,j), sminus(k,j), wplus(k,j), wminus(k,j), isdisc(A(l,j)), isdisc(j), fam(l,j),&
          parammat1(l,j), parammat2(l,j), tmp)
          nllk = nllk - tmp
        end do
      end do
      !call dblepr("levell", 8, nllk, 1)
      do k = 1,n
        if (.not. censorstatus(k)) then
          call mixedcopulallk(splus(k,d), sminus(k,d), wplus(k,d), wminus(k,d), isdisc(A(l,d)), isdisc(d), fam(1,d),&
          parammat1(l,d), parammat2(l,d), tmp)
          nllk = nllk - tmp
        end if
      end do
      !call dblepr("levell", 8, nllk, 1)
    else
      do k = 1,n
        if (.not. censorstatus(k)) then
          call mixedcopulallk(splus(k,d), sminus(k,d), wplus(k,d), wminus(k,d), isdisc(A(d-1,d)), isdisc(d), fam(d-1,d),&
          parammat1(d-1,d), parammat2(d-1,d), tmp)
          nllk = nllk - tmp
        end if
      end do
      !call dblepr("levell", 8, nllk, 1)
      if (isdisc(A(d-1, d))) then
        do k = 1,n
          if (censorstatus(k)) then
            call bivcopcdf(splus(k,d), wplus(k,d), fam(d-1,d), parammat1(d-1,d), parammat2(d-1,d), tmp1)
            call bivcopcdf(sminus(k,d), wplus(k,d), fam(d-1,d), parammat1(d-1,d), parammat2(d-1,d), tmp2)
            nllk = nllk - log(1 - (tmp1 - tmp2) / (splus(k,d) - sminus(k,d)))
          end if
        end do
        !call dblepr("levell", 8, nllk, 1)
      else
        do k = 1,n
          if (censorstatus(k)) then
            call bivcopccdf(wplus(k,d), splus(k,d), fam(d-1,d), parammat1(d-1,d), parammat2(d-1,d), tmp)
            nllk = nllk - log(1 - tmp)
          end if
        end do
        !call dblepr("levell", 8, nllk, 1)
      end if
    end if
  end do
  return
end

File: SurvCop/R/survivalnllk.r
#' negative log-likelihood of vine copula model for survival response
#' 
#' @description
#' negative log-likelihood of a vine copula model for a survival response (the last column of the data matrix) and mixed discrete/continuous explanatory variables (the other columns of the data matrix)
#' 
#' @param paramvec vector of length (d-1) if all the copula families associated with the survival response variable are one-parameter families; vector of length (d-1)+ otherwise
#' @param param1_complete d-1 by d-1 matrix for the first parameter of the copula models for the predictors
#' @param param2_complete d-1 by d-1 matrix for the second parameter of the copula models for the predictors
#' @param fam d by d matrix for the copula families coded by the VineCopula package
#' @param A d by d matrix for the copula structure
#' @param data_cdf_plus n by (d-1) matrix for CDF from the right
#' @param data_cdf_minus n by (d-1) matrix for CDF from the left
#' @param censor_status vector of length n representing if the response variable is censored or not for each observation
#' @param is_disc vector of length d representing if each variable is discrete or not
#'
#' @return the negative log-likelihood of the specified vine copula model for a survival response
#'
#' @import VineCopula
#' @export

survivalnllk = function(paramvec, param1_complete, param2_complete, fam, A, data_cdf_plus, data_cdf_minus, censor_status, is_disc) {
	d = ncol(A)
	n = nrow(data_cdf_plus)
	
	# permute all arguments based on the order of A
	A_order = diag(A)
	A = varrayperm(A, order(A_order))
	
	data_cdf_plus = data_cdf_plus[, A_order]
	data_cdf_minus = data_cdf_minus[, A_order]
	is_disc = is_disc[A_order]
	
	paramvec = parse_parameter(paramvec, fam[1:(d-1), d], d)
	param1_vec = paramvec[1:(d-1)]
	param2_vec = paramvec[d:length(paramvec)]
	
	parammat1 = matrix(0, nrow = d, ncol = d)
	parammat1[1:(d-1), 1:(d-1)] = param1_complete
	parammat1[1:(d-1), d] = param1_vec
	
	parammat2 = matrix(0, nrow = d, ncol = d)
	parammat2[1:(d-1), 1:(d-1)] = param2_complete
	parammat2[1:(d-1), d] = param2_vec
	
	v_plus = matrix(0, nrow = n, ncol = d)
	v_minus = matrix(0, nrow = n, ncol = d)
	vp_plus = matrix(0, nrow = n, ncol = d)
	vp_minus = matrix(0, nrow = n, ncol = d)
	s_plus = matrix(0, nrow = n, ncol = d)
	s_minus = matrix(0, nrow = n, ncol = d)
	w_plus = matrix(0, nrow = n, ncol = d)
	w_minus = matrix(0, nrow = n, ncol = d)
	
	M = matrix(0, nrow = d, ncol = d)
	for (i in 1:d) {
		for (j in i:d) {
			M[i, j] = max(A[1:i, j])
		}
	}
	
	I = matrix(0, nrow = d, ncol = d)
	for (i in 2:d) {
		for (j in i:d) {
			if (A[i, j] < M[i, j]) {
				I[i-1, M[i, j]] = 1
			}
		}
	}
	
	nllk = 0
	
	# initialization steps
	for (j in 1:d) {
		s_plus[, j] = data_cdf_plus[, A[1, j]]
		s_minus[, j] = data_cdf_minus[, A[1, j]]
		w_plus[, j] = data_cdf_plus[, j]
		w_minus[, j] = data_cdf_minus[, j]
	}
	
	# first level
	# uncensored
	for (j in 2:(d-1)) {
		nllk = nllk - sum(mixed_copula_llk(s_plus[, j], s_minus[, j], w_plus[, j], w_minus[, j], is_disc[A[1, j]], is_disc[j], fam[1, j], parammat1[1, j], parammat2[1, j]))
	}
	#print(nllk)
	# censored
	nllk = nllk - sum(mixed_copula_llk(s_plus[!censor_status, d], s_minus[!censor_status, d], w_plus[!censor_status, d], w_minus[!censor_status, d], is_disc[A[1, d]], is_disc[d], fam[1, d], parammat1[1, d], parammat2[1, d]))
	#print(nllk)
	
	# levels 2:(d-1)
	for (l in 2:(d-1)) {
		s_plus[s_plus >= .999] = .999
		s_minus[s_minus >= .999] = .998
		w_plus[w_plus >= .999] = .999
		w_minus[w_minus >= .999] = .998
		s_plus[s_plus <= .001] = .002
		s_minus[s_minus <= .001] = .001
		w_plus[w_plus <= .001] = .002
		w_minus[w_minus <= .001] = .001
		for (j in l:d) {
			if (I[l-1, j] == 1) {
				if (is_disc[j]) {
					vp_plus[, j] = (BiCopCDF(s_plus[, j], w_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_plus[, j], w_minus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (w_plus[, j] - w_minus[, j])
					vp_minus[, j] = (BiCopCDF(s_minus[, j], w_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_minus[, j], w_minus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (w_plus[, j] - w_minus[, j])
				} else {
					# condition on second variable for vp=vprime
					vp_plus[, j] = BiCopHfunc2(s_plus[, j], w_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
					vp_minus[, j] = BiCopHfunc2(s_minus[, j], w_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
				}
			}
			if (is_disc[A[l-1, j]]) {
				v_plus[, j] = (BiCopCDF(s_plus[, j], w_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_minus[, j], w_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (s_plus[, j] - s_minus[, j])
				v_minus[, j] = (BiCopCDF(s_plus[, j], w_minus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j]) - BiCopCDF(s_minus[, j], w_minus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])) / (s_plus[, j] - s_minus[, j])
			} else {
				# condition on first variable for v
				v_plus[, j] = BiCopHfunc2(w_plus[, j], s_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
				v_minus[, j] = BiCopHfunc2(w_minus[, j], s_plus[, j], fam[l-1, j], parammat1[l-1, j], parammat2[l-1, j])
			}
		}
		
		for (j in (l+1):d) {
			if (A[l, j] == M[l, j]) {
				s_plus[, j] = v_plus[, M[l, j]]
				s_minus[, j] = v_minus[, M[l, j]]
			} else {
				s_plus[, j] = vp_plus[, M[l, j]]
				s_minus[, j] = vp_minus[, M[l, j]]
			}
			w_plus[, j] = v_plus[, j]
			w_minus[, j] = v_minus[, j]
		}
		
		if (l < (d - 1)) {
			# uncensored
			for (j in (l+1):(d-1)) {
				nllk = nllk - sum(mixed_copula_llk(s_plus[, j], s_minus[, j], w_plus[, j], w_minus[, j], is_disc[A[l, j]], is_disc[j], fam[l, j], parammat1[l, j], parammat2[l, j]))
			}
			#print(nllk)
			# censored
			nllk = nllk - sum(mixed_copula_llk(s_plus[!censor_status, d], s_minus[!censor_status, d], w_plus[!censor_status, d], w_minus[!censor_status, d], is_disc[A[l, d]], is_disc[d], fam[l, d], parammat1[l, d], parammat2[l, d]))
			#print(nllk)
		} else {
			# uncensored
			nllk = nllk - sum(mixed_copula_llk(s_plus[!censor_status, d], s_minus[!censor_status, d], w_plus[!censor_status, d], w_minus[!censor_status, d], is_disc[A[d-1, d]], is_disc[d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d]))
			#print(nllk)
			if (is_disc[A[d-1, d]] & sum(censor_status) > 0) {
				# censored
				nllk = nllk - sum(log(1 - (BiCopCDF(s_plus[censor_status, d], w_plus[censor_status, d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d]) - BiCopCDF(s_minus[censor_status, d], w_plus[censor_status, d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d])) / (s_plus[censor_status, d] - s_minus[censor_status, d])))
				#print(nllk)
			} else {
				# censored
				nllk = nllk - sum(log(1 - BiCopHfunc2(w_plus[censor_status, d], s_plus[censor_status, d], fam[d-1, d], parammat1[d-1, d], parammat2[d-1, d])))
				#print(nllk)
			}
		}
	}
	return(nllk)
}